torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -0,0 +1,117 @@
+#include "torch_xla/csrc/xla_lower_util.h"
+
+#include <algorithm>
+#include <vector>
+
+#include "tensorflow/compiler/xla/shape_util.h"
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "tensorflow/compiler/xla/xla_client/util.h"
+#include "torch_xla/csrc/helpers.h"
+
+namespace torch_xla {
+namespace {
+
+std::pair<xla::XlaOp, xla::Shape> DotExpand(const xla::XlaOp& op,
+                                            const xla::Shape& op_shape,
+                                            const xla::Shape& to_shape) {
+  xla::int64 rank_delta = to_shape.rank() - op_shape.rank();
+  XLA_CHECK_GT(rank_delta, 0);
+
+  std::vector<xla::int64> reshape_sizes(to_shape.rank(), 1);
+  std::copy(op_shape.dimensions().begin(), op_shape.dimensions().end(),
+            reshape_sizes.begin() + rank_delta);
+  xla::XlaOp result = xla::Reshape(op, reshape_sizes);
+
+  std::vector<xla::int64> bcast_sizes(
+      to_shape.dimensions().begin(),
+      to_shape.dimensions().begin() + rank_delta);
+  bcast_sizes.insert(bcast_sizes.end(), op_shape.dimensions().begin(),
+                     op_shape.dimensions().end());
+  return std::make_pair(
+      xla::BroadcastInDim(result, bcast_sizes,
+                          xla::util::Iota<xla::int64>(to_shape.rank())),
+      xla::ShapeUtil::MakeShape(op_shape.element_type(), bcast_sizes));
+}
+
+std::pair<xla::XlaOp, xla::XlaOp> DotBroadcast(const xla::XlaOp& lhs,
+                                               const xla::Shape& lhs_shape,
+                                               const xla::XlaOp& rhs,
+                                               const xla::Shape& rhs_shape) {
+  auto lhs_dimensions = lhs_shape.dimensions();
+  auto rhs_dimensions = rhs_shape.dimensions();
+  XLA_CHECK_EQ(lhs_dimensions.size(), rhs_dimensions.size());
+  for (xla::int64 i = 0; i < lhs_dimensions.size() - 2; ++i) {
+    if (lhs_dimensions[i] == rhs_dimensions[i]) {
+      continue;
+    }
+    if (lhs_dimensions[i] == 1) {
+      lhs_dimensions[i] = rhs_dimensions[i];
+    } else if (rhs_dimensions[i] == 1) {
+      rhs_dimensions[i] = lhs_dimensions[i];
+    } else {
+      XLA_ERROR() << "Unsupported DotBroadcast: " << lhs_shape << " vs. "
+                  << rhs_shape;
+    }
+  }
+
+  xla::XlaOp blhs = lhs;
```

Comment:
`broadcasted_lhs`, `broadcasted_rhs` is not so bad, the right hand side expressions go to the next line anyway.

Reply:
SG